Events with zero-length summary rejected

If I put an event like this into Redis:
```
{
    "entity": "web-lb-gen-001.us-east-1.aws.prd.myco.com",
    "check": "serfHealth",
    "type": "service", 
    "state": "ok", 
    "summary": "",
    "details": "",
    "time": 1400171748
}
```

I get this error, and the event is dropped:
```
2014-05-15T16:36:29+00:00 [ERROR] :: flapjack-processor :: Invalid event data received, Event hash has missing keys 'summary' {"entity"=>"web-lb-gen-001.us-east-1.aws.prd.myco.com", "check"=>"serfHealth", "type"=>"service", "state"=>"ok", "summary"=>"", "details"=>"", "time"=>1400171748}
```

The `summary` key *does* exist, but it's an empty string.

--- a/lib/flapjack/data/event.rb
+++ b/lib/flapjack/data/event.rb
@@ -13,9 +13,10 @@ module Flapjack
       attr_reader :check, :summary, :details, :acknowledgement_id, :perfdata,
                   :initial_failure_delay, :repeat_failure_delay
 
-      REQUIRED_KEYS = ['type', 'state', 'entity', 'check', 'summary']
-      OPTIONAL_KEYS = ['time', 'details', 'acknowledgement_id', 'duration', 'tags', 'perfdata',
-                       'initial_failure_delay', 'repeat_failure_delay']
+      REQUIRED_KEYS = ['type', 'state', 'entity', 'check']
+      OPTIONAL_KEYS = ['time', 'summary', 'details', 'acknowledgement_id',
+                       'duration', 'tags', 'perfdata', 'initial_failure_delay',
+                       'repeat_failure_delay']
 
       VALIDATIONS = {
         proc {|e| e['type'].is_a?(String) &&
@@ -34,9 +35,6 @@ module Flapjack
         proc {|e| e['check'].is_a?(String) } =>
           "check must be a string",
 
-        proc {|e| e['summary'].is_a?(String) } =>
-          "summary must be a string",
-
         proc {|e| e['time'].nil? ||
                   e['time'].is_a?(Integer) ||
                  (e['time'].is_a?(String) && !!(e['time'] =~ /^\d+$/)) } =>
@@ -51,6 +49,9 @@ module Flapjack
                   e['repeat_failure_delay'].is_a?(Integer) ||
                  (e['repeat_failure_delay'].is_a?(String) && !!(e['repeat_failure_delay'] =~ /^\d+$/)) } =>
           "repeat_failure_delay must be a positive integer, or a string castable to one",
+
+        proc {|e| e['summary'].nil? || e['summary'].is_a?(String) } =>
+          "summary must be a string",
 
         proc {|e| e['details'].nil? || e['details'].is_a?(String) } =>
           "details must be a string",
@@ -234,6 +235,8 @@ module Flapjack
          'repeat_failure_delay'].each do |key|
           instance_variable_set("@#{key}", attrs[key])
         end
+        # summary is optional. set it to nil if it only contains whitespace
+        @summary = (@summary.is_a?(String) && ! @summary.strip.empty?) ? @summary.strip : nil
         # details is optional. set it to nil if it only contains whitespace
         @details = (@details.is_a?(String) && ! @details.strip.empty?) ? @details.strip : nil
         # perfdata is optional. set it to nil if it only contains whitespace

--- a/lib/flapjack/gateways/aws_sns/alert.text.erb
+++ b/lib/flapjack/gateways/aws_sns/alert.text.erb
@@ -1,5 +1,6 @@
+<% summary = @alert.summary %>
 <%= @alert.type_sentence_case %>: '<%= @alert.check %>' on <%= @alert.entity -%>
 <% unless ['acknowledgement', 'test'].include?(@alert.notification_type) -%>
  is <%= @alert.state_title_case -%>
 <% end -%>
- at <%= Time.at(@alert.time).strftime('%-d %b %H:%M') %>, <%= @alert.summary -%>
+ at <%= Time.at(@alert.time).strftime('%-d %b %H:%M') %><%= (summary.nil? || summary.empty?) ? '' : ", #{summary}" -%>

--- a/lib/flapjack/gateways/email/alert.html.erb
+++ b/lib/flapjack/gateways/email/alert.html.erb
@@ -32,10 +32,12 @@
       <td><%= @alert.state_title_case %></td>
     </tr>
 
-    <tr>
-      <td><strong>Summary</strong></td>
-      <td><%= @alert.summary %></td>
-    </tr>
+    <% if @alert.summary %>
+      <tr>
+        <td><strong>Summary</strong></td>
+        <td><%= @alert.summary %></td>
+      </tr>
+    <% end %>
 
     <% if @alert.details %>
       <tr>

--- a/lib/flapjack/gateways/email/alert.text.erb
+++ b/lib/flapjack/gateways/email/alert.text.erb
@@ -5,7 +5,9 @@ Monitoring has detected the following:
 Entity:           <%= @alert.entity %>
 Check:            <%= @alert.check %>
 State:            <%= @alert.state_title_case %>
+<% if @alert.summary -%>
 Summary:          <%= @alert.summary %>
+<% end -%>
 <% if @alert.details -%>
 Details:          <%= @alert.details %>
 <% end -%>

--- a/lib/flapjack/gateways/jabber/alert.text.erb
+++ b/lib/flapjack/gateways/jabber/alert.text.erb
@@ -7,6 +7,6 @@
 <% if ['acknowledgement'].include?(@alert.type) -%>
  has been acknowledged, unscheduled maintenance created for <%= time_period_in_words(@alert.acknowledgement_duration) -%>
 <% end -%>
-<% if @alert.summary && @alert.summary.length > 0 -%>
+<% if @alert.summary && !@alert.summary.empty? -%>
 , <%= @alert.summary -%>
 <% end -%>

--- a/lib/flapjack/gateways/pagerduty/alert.text.erb
+++ b/lib/flapjack/gateways/pagerduty/alert.text.erb
@@ -5,6 +5,6 @@
 <% if ['acknowledgement'].include?(@alert.type) -%>
  has been acknowledged, unscheduled maintenance created for <%= time_period_in_words(@alert.acknowledgement_duration) -%>
 <% end -%>
-<% if @alert.summary && @alert.summary.length > 0 -%>
+<% if @alert.summary && !@alert.summary.empty? -%>
 , <%= @alert.summary -%>
 <% end -%>

--- a/lib/flapjack/gateways/sms_messagenet/alert.text.erb
+++ b/lib/flapjack/gateways/sms_messagenet/alert.text.erb
@@ -1,5 +1,6 @@
+<% summary = @alert.summary -%>
 <%= @alert.type_sentence_case %>: '<%= @alert.check %>' on <%= @alert.entity -%>
 <% unless ['acknowledgement', 'test'].include?(@alert.notification_type) -%>
  is <%= @alert.state_title_case -%>
 <% end -%>
- at <%= Time.at(@alert.time).strftime('%-d %b %H:%M') %>, <%= @alert.summary -%>
+ at <%= Time.at(@alert.time).strftime('%-d %b %H:%M') %><%= (summary.nil? || summary.empty?) ? '' : ", #{summary}" -%>

--- a/lib/flapjack/gateways/sms_twilio/alert.text.erb
+++ b/lib/flapjack/gateways/sms_twilio/alert.text.erb
@@ -1,5 +1,6 @@
+<% summary = @alert.summary -%>
 <%= @alert.type_sentence_case %>: '<%= @alert.check %>' on <%= @alert.entity -%>
 <% unless ['acknowledgement', 'test'].include?(@alert.notification_type) -%>
  is <%= @alert.state_title_case -%>
 <% end -%>
- at <%= Time.at(@alert.time).strftime('%-d %b %H:%M') %>, <%= @alert.summary -%>
+ at <%= Time.at(@alert.time).strftime('%-d %b %H:%M') %><%= (summary.nil? || summary.empty?) ? '' : ", #{summary}" -%>

--- a/spec/lib/flapjack/data/event_spec.rb
+++ b/spec/lib/flapjack/data/event_spec.rb
@@ -101,7 +101,7 @@ describe Flapjack::Data::Event do
       expect(result).to be_nil
     end
 
-    ['type', 'state', 'entity', 'check', 'summary'].each do |required_key|
+    ['type', 'state', 'entity', 'check'].each do |required_key|
 
       it "rejects an event with missing '#{required_key}' key (archiving)" do
         bad_event_data = event_data.clone
@@ -168,7 +168,7 @@ describe Flapjack::Data::Event do
       end
     end
 
-    ['time', 'details', 'perfdata', 'acknowledgement_id', 'duration', 'tags'].each do |optional_key|
+    ['time', 'summary', 'details', 'perfdata', 'acknowledgement_id', 'duration', 'tags'].each do |optional_key|
       it "rejects an event with invalid '#{optional_key}' key (archiving)" do
         bad_event_data = event_data.clone
         bad_event_data[optional_key] = {'hello' => 'there'}

